postgresql: fix cross for FreeBSD

rhelmot opened the pull request:
- Fix dependencies
- Fix build of plugins in a cross context by providing pg_config_native copied from an equivalent native build and wrapped to provide the correct paths. The postgres source code indicates that there have been plans to do this upstream for a while, but that they are too complex to do "correctly".
- Fix the --export-dynamic flag - it is applicable for all lld linkers, and it should be spelled differently.

Pull requests using pg_config_native forthcoming.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-freebsd (cross)
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rhelmot committed:
postgresql: fix cross for FreeBSD

- Fix dependencies (uuid and locale)
- Add the --export-dynamic flag - it is applicable for all
  autotools-based cross configurations starting with v16.

drupol commented:
It should target `staging` too.

rhelmot commented:
Hm... darwin mass rebuild makes sense but I was really expecting rebuild-linux: 0.

rhelmot commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -377,6 +387,20 @@ let
             --replace-fail '-bundle_loader $(bindir)/postgres' "-bundle_loader $out/bin/postgres"
+        ''
+        + lib.optionalString (stdenv'.hostPlatform != stdenv.buildPlatform) ''
+          cp ${nativeBuild.dev}/bin/.pg_config-wrapped $dev/bin/.pg_config_native-wrapped
+          # wrapProgram will use the wrong shell. do it ourselves
+          cat >"$dev/bin/pg_config_native" <<EOF
+          #!${lib.getExe buildPackages.bash}
+          exec > >(sed ${
+            lib.concatMapStringsSep " " (
+              output: "-e s@${nativeBuild."${output}"}@${builtins.placeholder output}@g"
+            ) finalAttrs.outputs
+          })
+          exec -a "${builtins.placeholder "out"}/bin/pg_config" "${builtins.placeholder "dev"}/bin/.pg_config_native-wrapped" "\$@"
+          EOF
+          chmod +x "$dev/bin/pg_config_native"
```
I agree, it is potentially very fragile. I don't really know how to approach the "correct" solution though, and this does seem to work while at the same time being simple enough that its intent (I can document this inline) is pretty basic and can be easily understood as a hack to be replaced as soon as it gives anyone trouble.

Do you think it's a bad enough hack that it shouldn't be in nixpkgs in order to build what it can?

rhelmot commented:
Fixed all the issues that don't require rewriting pg_config.

rhelmot commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -127,6 +129,9 @@ let
 
       pg_config = writeShellScriptBin "pg_config" (builtins.readFile ./pg_config.sh);
+
+      nativeBuild =
+        buildPackages."postgresql_${lib.versions.major version}${lib.optionalString jitSupport "_jit"}";
```
Indeed. This approach is predicated on the assumption that the paths are the only thing that are essential. So far this is correct, but critically, I have been only compiling pgrx (rust) plugins.

rhelmot commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -277,6 +282,14 @@ let
         ++ lib.optionals (stdenv'.hostPlatform.isDarwin && atLeast "16" && olderThan "18") [
           "LDFLAGS_EX_BE=-Wl,-export_dynamic"
         ]
+        # some version of this flag is required in all cross configurations
+        # since it cannot be automatically detected
+        ++
+          lib.optionals
+            ((!stdenv'.hostPlatform.isDarwin) && (!(stdenv'.buildPlatform.canExecute stdenv.hostPlatform)))
+            [
+              "LDFLAGS_EX_BE=-Wl,--export-dynamic"
+            ]
```
I did this test with postgresql_16 just now (with only the uuid=bsd patch) and it was missing the symbols, so it sounds like we might be able to backport some postgresql_17 patch for this.

rhelmot replied:
okay, something really really weird is going on. I can't reproduce your results with current staging (ac65f8db981dba7611ac54b03bd1f60a72a2b748) and the following patch applied:

```diff
diff --git a/pkgs/servers/sql/postgresql/generic.nix b/pkgs/servers/sql/postgresql/generic.nix
index ca157411536e..6aaf4fb76f7e 100644                                                      
--- a/pkgs/servers/sql/postgresql/generic.nix                                                
+++ b/pkgs/servers/sql/postgresql/generic.nix                                                
@@ -19,6 +19,7 @@ let                                                                        
                                                                                             
       # runtime dependencies                                                                
       darwin,                                                                               
+      freebsd,                                                                              
       glibc,                                                                                
       libuuid,                                                                              
       libxml2,                                                                              
@@ -265,7 +266,7 @@ let                                                                      
           "--with-system-tzdata=${tzdata}/share/zoneinfo"                                   
           "--enable-debug"                                                                  
           (lib.optionalString systemdSupport "--with-systemd")                              
-          "--with-uuid=e2fs"                                                                
+          (if stdenv.hostPlatform.isFreeBSD then "--with-uuid=bsd" else "--with-uuid=e2fs") 
         ]                                                                                   
         ++ lib.optionals lz4Enabled [ "--with-lz4" ]                                        
         ++ lib.optionals zstdEnabled [ "--with-zstd" ]                                      
@@ -304,7 +305,7 @@ let                                                                      
                                                                                             
           (replaceVars ./patches/locale-binary-path.patch {                                 
             locale = "${                                                                    
-              if stdenv.hostPlatform.isDarwin then darwin.adv_cmds else lib.getBin stdenv.cc.libc
+              if stdenv.hostPlatform.isDarwin then darwin.adv_cmds else if stdenv.hostPlatform.isFreeBSD then freebsd.locale else lib.getBin stdenv.cc.libc
             }/bin/locale";                                                                  
           })                                                                                
         ]
```

```
nm /nix/store/qacy837wb0d5a0kjq2n3f6qwpr7sg3xz-postgresql-x86_64-unknown-freebsd-17.4/bin/postgres | grep lowerstr
0000000000a064d0 t lowerstr_with_len
```

(lowerstr_with_len is NOT the symbol we're looking for)

I'm compiling from x86_64-linux, with sandbox. Not sure what else could nail down this issue.

rhelmot replied:
The symbol seems to be present in postgresql 13-15, but not 16 or 17.

rhelmot commented:
I've adjusted the diff as requested. I'll look at pkgsStatic right now!

wolfgangwalther reviewed (approved):
All five major versions cross-compile fine for me. `libpq` does so already anyway.

rhelmot commented:
pkgsStatic status: Fixed openssl, fixed tzdata, but now libpq is being problematic because it always tries to build both static and dynamic executables, and the freebsd pkgsStatic stdenv just straight up cannot do dynamic stuff. I think I am going to leave that one to rest for now.

Will merge this if no more objections in the next several hours.
